src/can.rs

```diff
@@ -0,0 +1,511 @@
+                    let upper = (rx.rir.read().stid().bits() as u32) << 18;
+                    CanId::ExtendedId(upper | lower)
+                }
+                false => CanId::BaseId(rx.rir.read().stid().bits()),
+            };
+
+            // Release the mailbox
+            can.rfr[self.idx].modify(|_, w| w.rfom().set_bit());
+
+            let frame = CanFrame::new_with_len(rcv_id, &data, len);
+
+            return Ok(frame);
+        }
+
+        Err(nb::Error::WouldBlock)
+    }
+
+    fn set_filter(&mut self, filter: Self::Filter) {
+        // TODO: this likely needs to be in a critical section, is that OK?
+
+        let can = unsafe { &*CAN::ptr() };
+
+        // Filter init mode
+        can.fmr.modify(|_, w| w.finit().set_bit());
+
+        can.fm1r.modify(|_, w| match filter.data {
+            CanFilterData::MaskFilter(_, _) => w.fbm0().clear_bit(),
+            CanFilterData::ExtendedMaskFilter(_, _) => w.fbm0().clear_bit(),
+            CanFilterData::AcceptAll => w.fbm0().clear_bit(),
+            CanFilterData::IdFilter(_) => w.fbm0().set_bit(),
+        });
+
+        can.fs1r.modify(|_, w| match filter.data {
+            CanFilterData::MaskFilter(_, _) => w.fsc0().clear_bit(),
+            CanFilterData::ExtendedMaskFilter(_, _) => w.fsc0().set_bit(),
+            CanFilterData::AcceptAll => w.fsc0().set_bit(),
+            CanFilterData::IdFilter(id) => match id {
+                CanId::BaseId(_) => w.fsc0().clear_bit(),
+                CanId::ExtendedId(_) => w.fsc0().set_bit(),
+            },
+        });
+
+        // Assign filter to this FIFO
+        can.ffa1r.modify(|_, w| match self.idx {
+            0 => w.ffa0().clear_bit(),
+            1 => w.ffa0().set_bit(),
+            _ => unreachable!(),
+        });
+
+        let index = filter
+            .index
+            .unwrap_or_else(|| FILTER_INDEX.fetch_add(1, Ordering::Acquire))
+            as usize;
+
+        assert!(index < 28, "Filter index out of range");
+
+        can.fb[index]
+            .fr1
+            .modify(|_, w| unsafe { w.bits(filter.data.fr1()) });
+
+        let fr2 = filter.data.fr2().unwrap_or(0);
+        can.fb[index].fr2.modify(|_, w| unsafe { w.bits(fr2) });
+
+        can.fa1r.modify(|_, w| {
+            // TODO: the rest of these
+            match index {
+                0 => w.fact0().set_bit(),
+                1 => w.fact1().set_bit(),
+                2 => w.fact2().set_bit(),
+                3 => w.fact3().set_bit(),
+                4 => w.fact4().set_bit(),
+                5 => w.fact5().set_bit(),
+                _ => unimplemented!(),
+            }
+        });
+
+        // Disable init mode
+        can.fmr.modify(|_, w| w.finit().clear_bit());
+    }
+
+    fn clear_filter(&mut self) {
+        self.set_filter(CanFilter::new(CanFilterData::AcceptAll))
+    }
+}
+
+impl CanFrame {
+    pub fn new_with_len(id: CanId, data: &[u8], length: usize) -> CanFrame {
```

Comment:
Just curious. Why must the data slice be converted into an array? Slice has a [`len()`](https://doc.rust-lang.org/std/primitive.slice.html#method.len) method. Maybe you could give me some insight :)

Reply:
I should likely limit the visibility of this method and it could be simplified a bit... The reason this exists is mainly to do with lifetimes and there are sort of two reasons for it:

1. is for reading, we allocate an array always of length 8 bytes (the max CAN frame size), because we cant do dynamic allocation and then use a subslice of it, but now that I think more about it, we can likely just consume that array in this case and avoid the reference.
2. Is because of how we need to implement `embedded_hal_can::Frame`. Having every frame have a 8 byte backing array was just the simplest way to handle this without getting into using `generic_array` or something, but ultimately I may do that, because it would make the code clearer.